kconfigfunctions: add dt_path_enabled, dt_alias_enabled

mbolivar-nordic opened the pull request:
This is just like dt_nodelabel_enabled, but for aliases and paths.

It is explicitly being added for the benefit of out of tree applications, just to be able to rely on it in Zephyr without having to maintain out of tree patches.

mbolivar-nordic committed:
kconfig: add dt_path_enabled, dt_alias_enabled

There is already a way to check if a node is enabled based on its node
label, but we don't have an equivalent way to do that for a path or an
alias. Add them. These rely on the same underlying edtlib machinery to
get their jobs done, but are being treated differently within Kconfig
in order to match distinctions between paths and aliases made in the
devicetree.h API.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

galak reviewed (changes requested):
This should be dt_path_enabled as you can pass a path or an alias to such a function.

mbolivar-nordic commented:
> This should be dt_path_enabled as you can pass a path or an alias to such a function.

This function only works on aliases: https://gist.github.com/mbolivar-nordic/7a9207a0202d64d28abed3a3e989000a

Are you saying you also want a dt_path_enabled() function in addition to dt_alias_enabled()?

galak commented:
> > 
> 
> This function only works on aliases: https://gist.github.com/mbolivar-nordic/7a9207a0202d64d28abed3a3e989000a
> 
> Are you saying you also want a dt_path_enabled() function in addition to dt_alias_enabled()?

What I'm saying is that alias is just a replacement for path.  So having just a dt_path_enabled() is sufficient and you can pass that a path or an alias.  I think EDT already takes care of allow for both.

mbolivar-nordic commented:
> What I'm saying is that alias is just a replacement for path. So having just a dt_path_enabled() is sufficient and you can pass that a path or an alias. I think EDT already takes care of allow for both.

@galak rebased and updated.

pabigot reviewed (changes requested):
This is contrary to @galak's request, but:

It seems a little weird that we to access a node in C with `DT_PATH()`, `DT_NODELABEL()`, `DT_ALIAS()`, `DT_INST()`, `DT_CHOSEN()`; but for Kconfig we have `dt_nodelabel_enabled()` (which is the same) and `dt_node_enabled()` (which corresponds to either `DT_PATH()` or `DT_ALIAS()`).

It would be much easier if the same terms were used in both places.  I would not expect `dt_path_enabled()` to take an alias; and there's no `DT_NODE()` that works whether it's given a path or an alias.

So I would like to see `dt_foo_enabled()` paired against `DT_FOO()` for all node identifiers, unless there's some reason that can't be done.

mbolivar-nordic commented:
@pabigot noted; I don't have a strong opinion here as long as the functionality is available. Will let you and @galak discuss; I'll update the PR when there is consensus.

galak commented:
> @pabigot noted; I don't have a strong opinion here as long as the functionality is available. Will let you and @galak discuss; I'll update the PR when there is consensus.

The reason I suggested dt_path_enabled is this mimics also the behavior of libfdt w/regards to it's "path" based APIs taking either a path or alias.

galak commented:
Dev-Review (Jul 9) implement both dt_path_enabled & dt_alias_enabled to be more explicit.  dt_path_enabled can take a path or alias, and dt_alias_enabled will only accept an alias.

mbolivar-nordic committed:
scripts: kconfigfunctions: add a comment

I can never remember what the contents of the kconfigfunctions dict is
and have to look it up every time. I suspect I'm not alone, and not
everyone who uses these will know how to look up the answer. Add a
comment explaining the value.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic commented:
Rebased and addressed feedback.

pabigot reviewed (approved):
This doesn't have `dt_inst_enabled()` so it's not an exact match for what we can do in C but we can go back and add that later if it's important (I still feel that it would be useful).

galak commented on scripts/kconfig/kconfigfunctions.py:
```diff
@@ -88,10 +88,40 @@ def dt_chosen_path(kconf, _, chosen):
+    This function is used to test if a node is enabled or not. If it's
+    called as "dt_path_enabled", it works with either aliases or
+    paths. If it is called as "dt_alias_enabled", it works only with
+    aliases.
```
Can we add to the comment about the 'node' argument being either the path or alias.

galak reviewed (commented):
LGTM, would like to see a bit more detail in the comment on args to dt_node_enabled

mbolivar-nordic commented:
> LGTM, would like to see a bit more detail in the comment on args to dt_node_enabled

done

mbolivar-nordic commented:
> This doesn't have `dt_inst_enabled()` so it's not an exact match for what we can do in C but we can go back and add that later if it's important (I still feel that it would be useful).

I'm not convinced about that. Individual instances are supposed to be indistinguishable so the only thing I could imagine being useful is how many instances are enabled, not which ones.

We already have dt_compat_enabled for testing whether the count of enabled instances is nonzero. I can't really think of a use for a "dt_num_instances_enabled" offhand, though I agree it could be added if there is a use case.
